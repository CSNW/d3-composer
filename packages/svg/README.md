# @d3-composer/svg

## Example

```js
import { chart, layer, lines, scatter } from '@d3-composer/svg';

function linesChart(selection, props) {
  const { width, height, data, xScale, yScale } = props;

  const svg = chart(selection, { width, height, responsive: true });

  lines(layer(svg, 'lines'), { data, xScale, yScale });
  scatter(layer(svg, 'scatter'), { data, xScale, yScale });
}
```

## API

<a href="#chart" name="chart">#</a> <b>chart</b>(<i>selection</i>[, <i>props</i>])

SVG chart wrapper with sizing and responsive options

<a href="#layout" name="layout">#</a> <b>layout</b>(<i>selection</i>[, <i>grid</i>])

Create `g` layers from grid

<a href="#layer" name="layer">#</a> <b>layer</b>(<i>selection</i>[, <i>id</i>])

Create named `g` layer

<a href="#lines" name="lines">#</a> <b>lines</b>(<i>selection</i>[, <i>props</i>])

Options:

- `data` - {x,y} series data
- `xScale` - [d3-scale](https://github.com/d3/d3-scale) for x-value
- `yScale` - d3-scale for y-value
- `[xValue]` - accessor for x-value (default = `d => d.x`)
- `[yValue]` - accessor for y-value (default = `d => d.y`)
- `[curve]` - See [d3-shape](https://github.com/d3/d3-shape#curves)
- `[style]` - Style string, object, or function for line
- `[class]` - Class string or function for line
- `[transition]` - An instance of `d3.transition()` (see [d3-transition](https://github.com/d3/d3-transition#selection_transition)).
- `[interpolate]` - An interpolation function for `update` and `exit`, such as [d3-interpolate-path](https://github.com/pbeshai/d3-interpolate-path)

```js
// scale
lines(selection, { xScale: d3.scaleLog().domain([1, 10000]) })

// curve
lines(selection, { curve: d3.curveCardinal() });

// style
lines(selection, { style: d => { stroke: d.color } });

// class
lines(selection, { class: d => d.class });

// transition
lines(selection, { transition: d3.transition().duration(1000) })

// interpolate
lines(selection, { interpolate: d3.interpolatePath })
```

<a href="#bars" name="bars">#</a> <b>bars</b>(<i>selection</i>[, <i>props</i>])

Options:

- `data` - {x,y} or {x0,x1,y} or {x0,x1,y0,y1} series data
- `xScale` - [d3-scale](https://github.com/d3/d3-scale) for x-value (`d3.scaleBand()` is the typical choice)
- `yScale` - d3-scale for y-value
- `[xValue]` - accessor for x-value (default = `d => d.x`)
- `[yValue]` - accessor for y-value (default = `d => d.y`)
- `[x0]` - accessor for x0-value (default is x0-position or x-position)
- `[x1]` - accessor for x1-value (default is x1-position or `x0 + xScale.bandwidth()`)
- `[y0]` - accessor for y0-value (default is y0-position or `yScale(0)`)
- `[y1]` - accessor for y1-value (default is y1-position or y-position)
- `[key]` - Bar key for identifying bars (default is `d => d.key or x-value`)
- `[style]` - Style string, object, or function for bar
- `[class]` - Class string or function for bar
- `[seriesKey]` - Series key for identifying series (default is `(series, i) => series.key || i`)
- `[seriesStyle]` - Style string, object, or function for series group
- `[seriesClass]` - Class string or function for series group
- `[transition]` - An instance of `d3.transition()` (see [d3-transition](https://github.com/d3/d3-transition#selection_transition)).

```js
// scale
bars(selection, { xScale: d3.scaleLog().domain([1, 10000]) })

// style
bars(selection, { style: d => { fill: d.color } });

// class
bars(selection, { class: d => d.class });

// transition
bars(selection, { transition: d3.transition().duration(1000) })
```

<a href="#scatter" name="scatter">#</a> <b>scatter</b>(<i>selection</i>[, <i>props</i>])

Scatter chart

<a href="#labels" name="labels">#</a> <b>labels</b>(<i>selection</i>[, <i>props</i>])

Labels component

<a href="#axis" name="axis">#</a> <b>axis</b>(<i>selection</i>[, <i>props</i>])

Axis component

<a href="#text" name="text">#</a> <b>text</b>(<i>selection</i>[, <i>props</i>])

Text component

<a href="#legend" name="legend">#</a> <b>legend</b>(<i>selection</i>[, <i>props</i>])

Legend component

<a href="#annotation" name="annotation">#</a> <b>annotation</b>(<i>selection</i>[, <i>props</i>])

Annotation component

<a href="#gridlines" name="gridlines">#</a> <b>gridlines</b>(<i>selection</i>[, <i>props</i>])

Gridlines component

<a href="#size" name="size">#</a> <b>size</b>(<i>selection<i>)

Size helper
