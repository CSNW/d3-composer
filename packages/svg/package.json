{
  "name": "@d3-composer/svg",
  "version": "0.3.1",
  "description": "svg components for d3-composer",
  "author": "Tim Hall <tim.hall.engr@gmail.com>",
  "license": "MIT",
  "repository": "https://github.com/CSNW/d3-composer.git",
  "contributors": [
    "CSNW"
  ],
  "bugs": {
    "url": "https://github.com/CSNW/d3-composer/issues"
  },
  "homepage": "https://github.com/CSNW/d3-composer",
  "keyword": [
    "d3",
    "d3-composer",
    "svg",
    "chart",
    "visualization"
  ],
  "main": "dist/d3-composer-svg.js",
  "unpkg": "dist/d3-composer-svg.min.js",
  "module": "src/index.js",
  "scripts": {
    "test": "exit 0",
    "build": "rollup -c"
  },
  "dependencies": {
    "@d3-composer/utils": "^0.3.1",
    "d3-axis": "^1",
    "d3-selection": "^1",
    "d3-shape": "^1"
  },
  "devDependencies": {
    "@d3-composer/grid": "^0.3.1"
  },
  "files": ["dist", "src", "!src/__tests__", "!src/__helpers__"]
}
