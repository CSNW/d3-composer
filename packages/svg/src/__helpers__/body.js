import { select } from 'd3';

export default function body() {
  let selection;
  let svg;
  const instances = [];

  return {
    setup() {
      selection = select(document.body);
      svg = selection.append('svg');
    },
    teardown() {
      for (const instance of instances) {
        instance.remove();
      }
      svg.remove();
    },

    append(type) {
      const instance = selection.append(type);
      instances.push(instance);

      return instance;
    },
    layer() {
      return svg.append('g');
    }
  };
}
