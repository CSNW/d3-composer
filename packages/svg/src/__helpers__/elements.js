const default_size = { width: 100, height: 100 };

export class SVGElement {
  constructor(options = {}) {
    const { size = default_size } = options;

    this.size = size;
  }

  getBBox() {
    return this.size;
  }
}
