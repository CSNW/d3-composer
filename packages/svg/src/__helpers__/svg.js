import { select } from 'd3';

export function svg() {
  let instance;

  function layer() {
    return instance.append('g');
  }

  layer.setup = () => {
    instance = select(document.body).append('svg');
  };
  layer.teardown = () => {
    instance.remove();
  };

  return layer;
}
