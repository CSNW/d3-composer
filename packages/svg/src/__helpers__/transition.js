export function end(transition) {
  return new Promise(resolve => {
    transition.on('end', () => resolve());
  });
}
