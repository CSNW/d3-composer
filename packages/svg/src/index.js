export { default as chart } from './chart';
export { default as layout } from './layout';
export { default as layer } from './layer';
export { default as series } from './series';
export { default as line } from './line';
export { default as bars } from './bars';
export { default as scatter } from './scatter';
export { default as area } from './area';
export { default as labels } from './labels';
export { axisLeft, axisRight, axisBottom, axisTop } from './axis';
export { default as text } from './text';
export { default as legend } from './legend';
export { default as gridlines } from './gridlines';
export { symbolLine } from './symbols';
export { size, interpolatePath, translateXY } from './utils';
