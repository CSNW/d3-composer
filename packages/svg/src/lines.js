import { xy, seriesKey } from '@d3-composer/utils';
import { line as d3_line } from 'd3-shape';

export default function lines(selection, props = {}) {
  const { curve, style } = props;
  const { data, x, y } = xy(props);

  const line = d3_line()
    .x(x)
    .y(y);
  if (curve) line.curve(curve);

  const lines = selection.selectAll('path').data(data, seriesKey);

  lines.exit().remove();

  lines
    .enter()
    .append('path')
    .merge(lines)
    .attr('d', d => line(d.values))
    .attr('style', style);

  return selection;
}
