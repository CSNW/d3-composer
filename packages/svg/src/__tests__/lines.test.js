import { scaleLinear, transition as d3_transition } from 'd3';
import { svg } from '../__helpers__/svg';
import { end } from '../__helpers__/transition';
import lines from '../lines';

const fixture = svg();

beforeAll(fixture.setup);
afterAll(fixture.teardown);

test('should render simple line', async () => {
  const transition = d3_transition().duration(0);
  const selection = lines(fixture(), {
    data: [{ x: 0, y: 0 }, { x: 100, y: 100 }],
    xScale: scaleLinear()
      .domain([0, 100])
      .range([0, 100]),
    yScale: scaleLinear()
      .domain([0, 100])
      .range([100, 0]),
    transition
  });

  await end(transition);

  expect(selection.node()).toMatchSnapshot();
});

test('should use style and class', async () => {
  const transition = d3_transition().duration(0);
  const selection = lines(fixture(), {
    data: [
      { class: 'a', values: [{ x: 0, y: 0 }, { x: 100, y: 100 }] },
      { class: 'b', values: [{ x: 0, y: 100 }, { x: 100, y: 0 }] }
    ],

    // TODO Make xScale and yScale optional
    xScale: scaleLinear()
      .domain([0, 100])
      .range([0, 100]),
    yScale: scaleLinear()
      .domain([0, 100])
      .range([100, 0]),
    transition,

    style: (_, i) => ({
      stroke: i % 2 === 0 ? 'blue' : 'red'
    }),
    class: d => d.class
  });

  await end(transition);

  expect(selection.node()).toMatchSnapshot();
});
