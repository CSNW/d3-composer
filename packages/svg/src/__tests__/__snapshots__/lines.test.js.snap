// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render simple line 1`] = `
<g>
  <path
    d="M0,100L100,0"
    fill="none"
  />
</g>
`;

exports[`should use style and class 1`] = `
<g>
  <path
    class="a"
    d="M0,100L100,0"
    fill="none"
    style="stroke: blue;"
  />
  <path
    class="b"
    d="M0,0L100,100"
    fill="none"
    style="stroke: red;"
  />
</g>
`;
